[ie/youtube] Add a PO Token Provider Framework (#12840)
yt-dlp/yt-dlp: README.md

* `player_js_variant`: The player javascript variant to use for signature and nsig deciphering. The known variants are: `main`, `tce`, `tv`, `tv_es6`, `phone`, `tablet`. Only `main` is recommended as a possible workaround; the others are for debugging purposes. The default is to use what is prescribed by the site, and can be selected with `actual`

##### PO Token settings
* `po_token`:  Proof of Origin (PO) Token(s) to use. Comma seperated list of PO Tokens in the format `CLIENT.CONTEXT+PO_TOKEN`, e.g. `youtube:po_token=web.gvs+XXX,web.player=XXX,web_safari.gvs+YYY`. Context can be either `gvs` (Google Video Server URLs) or `player` (Innertube player request)
* `pot_trace`: Enable debug logging for PO Token fetching. Either `true` or `false` (default)
* `fetch_pot`: Policy to use for fetching a PO Token from providers. `always` to always try fetch a PO Token regardless if the client requires one for the given context, `auto` to only fetch a PO Token if the client requires one for the given context (default)

###### youtubepot-webpo

> * `bind_to_visitor_id`: Whether to use the Visitor ID instead of Visitor Data for caching WebPO tokens. Either `true` or `false` (default `true`)

Author: moved to same level as other options, thanks

* `po_token`:  Proof of Origin (PO) Token(s) to use. Comma seperated list of PO Tokens in the format `CLIENT.CONTEXT+PO_TOKEN`, e.g. `youtube:po_token=web.gvs+XXX,web.player=XXX,web_safari.gvs+YYY`. Context can be either `gvs` (Google Video Server URLs) or `player` (Innertube player request)
* `player_js_variant`: The player javascript variant to use for signature and nsig deciphering. The known variants are: `main`, `tce`, `tv`, `tv_es6`, `phone`, `tablet`. Only `main` is recommended as a possible workaround; the others are for debugging purposes. The default is to use what is prescribed by the site, and can be selected with `actual`

##### PO Token settings
* `po_token`:  Proof of Origin (PO) Token(s) to use. Comma seperated list of PO Tokens in the format `CLIENT.CONTEXT+PO_TOKEN`, e.g. `youtube:po_token=web.gvs+XXX,web.player=XXX,web_safari.gvs+YYY`. Context can be either `gvs` (Google Video Server URLs) or `player` (Innertube player request)

> * `pot_debug`: Enable debug logging for PO Token fetching. Either `true` or `false` (default `false`)

Reviewer: Wouldn't debug levels (0,1,2 etc) be more useful? `trace` could have a level passed to it as arg. Also, When trace is on, we should force `debug` for POT even without `-v`

Author: Hmm, so as to say somewhat decouple it from YDL logger, but have -v also change the log level of it?

Author: I'll add a log level for the pot logger. However it is a subset of the YDL/IE logger level, so even if you enable POT debug, if you don't pass -v it will not show.

Author: I've reverted this extractor arg back to a true/false option as I don't think there is much use for the different log levels given its behind -v and it's only for debugging. However I've kept underlying implementation based on the levels.

* `fetch_pot`: Policy to use for fetching a PO Token from providers. `always` to always try fetch a PO Token regardless if the client requires one for the given context. `when_required` to only fetch a PO Token if the client requires one for the given context (default)

###### youtubepot-memory

> * `max_size`: Max size of the PO Token LRU memory cache (default 25)

Reviewer 2: What's maximum size measured in? This seems ambiguous

Author: Maximum items in the cache

Reviewer: ```suggestion
* `max_size`: Number of PO Tokens to save in LRU memory cache (default 25)
```

Do we even need to expose this, though?

Author: > He's right. It's confusing since cache is usually given in MB.
> 
> ```suggestion
> * `max_size`: Number of PO Tokens to save in LRU memory cache (default 25)
> ```
> 
> Do we even need to expose this, though?

Yeah thinking about it, we probably don't need to expose it, I don't see a need since it's a temp cache. Might be something we can add later if requested.

Author: > Shouldn't yt-dlp just cache one Po Token per client, since afaik a new Po Token will only be fetched if the old one is invalidated or a different client? Or are there other cases where a new Po Token will be fetched (such as a different IP address)?
> 
> EDIT: On a side note, perhaps `tokens_cached` would be a better name for that argument?

Not quite how PO Tokens work. You want the cache by what the token is bound to (which is not necessary by client). See the webpo cache spec which describes how they are cached for web clients.

##### PO Token settings
* `po_token`:  Proof of Origin (PO) Token(s) to use. Comma seperated list of PO Tokens in the format `CLIENT.CONTEXT+PO_TOKEN`, e.g. `youtube:po_token=web.gvs+XXX,web.player=XXX,web_safari.gvs+YYY`. Context can be either `gvs` (Google Video Server URLs) or `player` (Innertube player request)
* `pot_trace`: Enable debug logging for PO Token fetching. Either `true` or `false` (default)

> * `fetch_pot`: Policy to use for fetching a PO Token from providers. `always` to always try fetch a PO Token regardless if the client requires one for the given context, `auto` to only fetch a PO Token if the client requires one for the given context (default)

Reviewer 3: For `fetch_pot` should we support a policy of `never`? Or is it expected that a user who wants that should pass `--plugins-dir none`? (and then what if they want to use their other plugins?)

Author: ah yeah we could support that

Author: added

###### youtubepot-webpo

> * `bind_to_visitor_id`: Whether to use the Visitor ID instead of Visitor Data for caching WebPO tokens. Either `true` or `false` (default `true`)

Reviewer 3: Without even looking at the code behind this yet (as will be the case for most of our users), I think this is very confusing. It's a subheading under `youtube`, but `youtubepot-webpo` looks like it should be the extractor value (e.g. `--extractor-arg "youtubepot-webpo:bind_to_visitor_id=true"`). We should at least have an example here to remove the ambiguity
